File: leetcode/202501/MinimumLengthofStringAfterOperations/MinimumLengthofStringAfterOperations.py
# -*- encoding: utf-8 -*-
"""
@Introduce  : 
@File       : MinimumLengthofStringAfterOperations.py
@Author     : ryrl
@Email      : ryrl970311@gmail.com
@Time       : 2025/01/13 18:02
@Description: leetcode 3223. Minimum Length of String After Operations
"""
# You are given a string `s`.
# You can perform the following process on `s` **any** number of times:
# Choose an index `i` in the string such that there is at least one character to the left of index `i` that is equal to s[i], and at least one character to the right that is also equal to s[i].
# Delete the closest character to the left of index `i` that is equal to `s[i]`.
# Delete the closest character to the right of index `i` that is equal to `s[i]`.
# Return the `minimum` length of the final string `s` that you can achieve.

# Example 1:
# Input: s = "abaacbcbb"
# Output: 5
# Explanation:
# We do the following operations:
# Choose index 2, then remove the characters at indices 0 and 3. The resulting string is s = "bacbcbb".
# Choose index 3, then remove the characters at indices 0 and 5. The resulting string is s = "acbcb".

# Example 2:
# Input: s = "aa"
# Output: 2
# Explanation:
# We cannot perform any operations, so we return the length of the original string.


class Solution:
    def minimumLength(self, s: str) -> int:
        from collections import Counter

        counts = Counter(s)
        count2Subtraction = 0
        for val in counts.values():

            if val % 2 == 1:
                count2Subtraction += val -1
            else:
                count2Subtraction += val - 2
        return len(s) - count2Subtraction
    
    def minimumLength_(self, s: str) -> int:
        
        count2Subtraction = 0
        for i in set(s):
            count = s.count(i)
            if count % 2 == 1:
                count2Subtraction += count - 1
            else:
                count2Subtraction += count - 2
        return len(s) - count2Subtraction
    
    def minimumLength__(self, s: str) -> int:
        
        res = 0
        for i in set(s):
            count = s.count(i)
            if count % 2 == 1:
                res += 1
            else:
                res += 2
        return res

File: leetcode/202501/MinimumLengthofStringAfterOperations/MinimumLengthofStringAfterOperations.rs
// @Introduce  : 
// @File       : MinimumLengthofStringAfterOperations.rs
// @Author     : ryrl
// @Email      : ryrl970311@gmail.com
// @Time       : 2025/01/13 22:38
// @Description:

impl Solution {
    pub fn minimum_length(s: String) -> i32 {
        use std::collections::HashSet;

        let mut res = 0;
        let unique_chars: HashSet<_> = s.chars().collect();

        for &c in &unique_chars {
            let count = s.chars().filter(|&x| x == c).count();
            if count % 2 == 1 {
                res += 1;
            } else {
                res += 2;
            }
        }
        res
    }

    pub fn minimum_length_(s: String) -> i32 {
        let length = s.len() as i32;
        let mut length2Subtraction = 0;
        let mut freq = vec![0; 26];

        for c in s.chars() {freq[(c as u8 - b'a') as usize] += 1;}
        
        for frequency in freq {
            if frequency % 2 != 0 {
                removable_len += frequency - 1;
            } else if frequency != 0 {
                removable_len += frequency - 2;
            }
        }
        
        length - removable_len
    }
}